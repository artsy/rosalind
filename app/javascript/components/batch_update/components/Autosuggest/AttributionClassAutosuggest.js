import React from 'react'
import PropTypes from 'prop-types'
import GenericAutosuggest from './GenericAutosuggest'
import { getSuggestionValue, renderSuggestion } from './helpers'

const attributionClasses = [
  { name: 'Editioned multiple', value: 'editioned multiple' },
  { name: 'Ephemera', value: 'ephemera' },
  { name: 'Limited edition', value: 'limited edition' },
  { name: 'Made-to-order', value: 'made-to-order' },
  { name: 'Non-editioned multiple', value: 'non-editioned multiple' },
  { name: 'Reproduction', value: 'reproduction' },
  { name: 'Unique', value: 'unique' },
]

const getSuggestions = value => {
  const inputValue = value.trim().toLowerCase()
  const inputLength = inputValue.length

  const results =
    inputLength === 0
      ? attributionClasses
      : attributionClasses.filter(attr =>
          attr.name.toLowerCase().includes(inputValue)
        )
  return Promise.resolve(results)
}

function AttributionClassAutosuggest(props) {
  return (
    <GenericAutosuggest
      id="attribution-class-autosuggest"
      placeholder={props.placeholder}
      fetchSuggestions={getSuggestions}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      selectSuggestion={attributionClass => {
        props.updateState('attributionClass', attributionClass)
      }}
      shouldRenderSuggestions={() => true}
    />
  )
}

AttributionClassAutosuggest.propTypes = {
  updateState: PropTypes.func,
  placeholder: PropTypes.string,
}

AttributionClassAutosuggest.defaultProps = {
  placeholder: 'Select an attribution class',
}

export { AttributionClassAutosuggest }
