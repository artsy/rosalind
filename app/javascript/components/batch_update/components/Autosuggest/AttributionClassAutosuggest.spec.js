import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { AttributionClassAutosuggest } from './AttributionClassAutosuggest'
import { mount } from 'enzyme'

const mockUpdater = jest.fn()

it('renders correctly', () => {
  const rendered = renderer.create(
    <AttributionClassAutosuggest updateState={mockUpdater} />
  )
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})

it('displays all options upon focus', () => {
  const wrapper = mount(
    <AttributionClassAutosuggest updateState={mockUpdater} />
  )
  expect(wrapper.find('li[role="option"]')).toHaveLength(0)

  const input = wrapper.find('input[type="text"]')
  input.simulate('focus')
  setImmediate(() => {
    // wait for next tick for promised matches
    wrapper.update()
    expect(wrapper.find('li[role="option"]')).toHaveLength(7)
  })
})

it('filters as you type', () => {
  const wrapper = mount(
    <AttributionClassAutosuggest updateState={mockUpdater} />
  )
  expect(wrapper.find('li[role="option"]')).toHaveLength(0)

  const input = wrapper.find('input[type="text"]')
  input.simulate('focus')
  input.simulate('change', { target: { value: 'editio' } })
  setImmediate(() => {
    // wait for next tick for promised matches
    wrapper.update()
    expect(wrapper.find('li[role="option"]')).toHaveLength(3)
  })
})
