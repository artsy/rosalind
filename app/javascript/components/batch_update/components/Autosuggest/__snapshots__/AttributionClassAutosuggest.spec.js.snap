// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="GenericAutosuggest"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="react-autowhatever-attribution-class-autosuggest"
    className="react-autosuggest__container"
    role="combobox"
  >
    <input
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-controls="react-autowhatever-attribution-class-autosuggest"
      autoComplete="off"
      className="react-autosuggest__input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Select an attribution class"
      type="text"
      value=""
    />
    <div
      className="react-autosuggest__suggestions-container"
      id="react-autowhatever-attribution-class-autosuggest"
      role="listbox"
    />
  </div>
</div>
`;
